### PR TITLE
Caffe models support

### PR DESCRIPTION
**Pull request by fchollet**

### CODE REVIEW

**Comment by fchollet** [comment]
@pranv I haven't really looked at it yet, just added your code and cleaned it up. Check out the changes I made, they should help you improve your coding style.

I'll do a full review in the next few days.

**Comment by pranv** [comment]
@fchollet
Great!
Learnt a few things..

**Comment by fchollet** [comment]
Started code-reviewing. I find the code quite difficult to understand, and I believe it could be considerably simplified :(

**Comment by pranv** [comment]
As far as I know, the layers are treated in the standard way. The complexity is present in getting the graph. Again, as far as I know, this is pretty much standard. There aren't other libraries that do this, hence it was hard. And I'm pretty sure that all steps are required, and I couldn't figure out a way to simplify them.
Is there a specific area, or if the entire thing unnecessarily complex?

**Comment by fchollet** [comment]
I haven't understood everything yet. But this part seems quite complicated and difficult to follow:

```python
network = make_network(layers, phase)  # obtain the nodes that make up the graph
if len(network) == 0:
    raise Exception('failed to construct network from the prototext')
network = acyclic(network)  # Convert it to be truly acyclic
network = merge_layer_blob(network)  # eliminate 'blobs', just have layers
reverse_network = reverse(network)  # reverse, to obtain the network_input
network_inputs = get_inputs(reverse_network)  # inputs of the network - 'in-order' is zero
network_outputs = get_outputs(network)  # outputs of the network - 'out-order' is zero
network = remove_label_paths(network, network_inputs, network_outputs)  # path from input to loss layers removed
reverse_network = reverse(network)  # stores the 'input' for each layer
```

Would there be any way to convert the Caffe prototext to a structure such as [this](https://github.com/fchollet/keras/blob/master/keras/layers/containers.py#L219-L225) in one pass?

I'll look into it.

Otherwise, I'm worried a bit about redundancy. Technically we already have code to build a model from a 'config' (though the config format is a bit different, but could be unified): https://github.com/fchollet/keras/blob/master/keras/utils/layer_utils.py#L18

**Comment by pranv** [comment]
The caffe definition of models is a bit weird. Technically, and according to the documentation, caffe accepts all Directed Acyclic Graphs (DAG). But in reality, several functions like `ReLU`, `Dropout` etc, can start and end at the same blob. This is because the computation can happen at the same memory location, ie blob. This causes cycles and naively using the `bottom` parameter of each layer could lead to a disaster.

At the same time, there are no specified input and output layers. Since it is a DAG, the nodes with 0 in-degree are input nodes. Similarly, the nodes with 0 out-degree are output nodes. This has to be found out. Other methods that use "one pass" cannot handle things like Siamese Networks - they place a constraint that the model is sequential. 

I create a graph as a dictionary. I'll find the node which maps to none, that is the output node. I'll reverse the network, find nodes which maps to none to get the start nodes. 

Now one more important aspect of caffe is that the same `data` layer is usually used to get both the actual data and the labels. This is an inherent non-sequentiality and keras doesn't do things this way. So that path, between `data` and the final loss layers has to be removed.

The `reverse_network ` stores all the previous nodes of a given node. Hence it is a **node -> [previous nodes]** map.

Hope that clarifies some things.

**Comment by pranv** [comment]
Another addition: All names are blobs.

**Comment by asampat3090** [comment]
I noticed the 'data' input isn't present when loading the caffemodel. Is that by design? I'm not able to get my results to match with caffe for the VGG 16 layer network (prototxt and caffemodel). It seems to be an issue with the input layer being 'conv1_1' instead of 'data'.

**Comment by fchollet** [comment]
@asampat3090 @pranv Any thoughts on how this could be addressed?

**Comment by pranv** [comment]
I think I had addressed this issue before. The input name isn't `'data'` since your prototext is in deploy mode. In deploy mode, the first layer is directly a processing layer and not `DATA`. And that layer's name has to be given as the input.

I've added 'model.inputs' and 'model.outputs' to help in this regard.
```python
model = convert.CaffeToKeras(
    prototext='cnn_params/VGG_ILSVRC_16_layers_deploy_features.prototxt',
    caffemodel='cnn_params/VGG_ILSVRC_16_layers.caffemodel') #phase irrelavent in deploy mode
graph = model('network') 
graph.compile('rmsprop', {model.outputs[0]: 'mse'})
features = graph.predict({model.inputs[0]: batch_images}, batch_size=1, verbose=1)
```
And does pycaffe run the code in caffe or in keras? This PR is about converting caffe models to keras, and running the network in keras.

And although this is a couple of lines more, which code seems simpler to you?
Consider the fact that we can't expect all keras users to know caffe entirely

**Comment by fchollet** [comment]
The big question is:

> I'm not able to get my results to match with caffe for the VGG 16 layer network (prototxt and caffemodel).

So, does this implementation produce a Keras model that is capable of reproducing the Caffe results? Or Not? If not, what should be done to address it?

**Comment by asampat3090** [comment]
Yea @pranv I understand the syntax but the concern is that it doesn't return the same result. As far as I can tell I have used the correct Keras code. Have you tried running this against caffe for known models?  Doesn't seem to match up. @fchollet I'll have to delve deeper into the code to figure out next steps.

**Comment by pranv** [comment]
@asampat3090 what exactly is the difference in result?

**Comment by asampat3090** [comment]
They just don't match up at all. What would be the best way to show you the difference? Can I send you the pickle files? For context, I am taking the output features from the 'fc7' layer w/ a 1x4096 feature vector.

**Comment by pranv** [comment]
@asampat3090 okay. Do go through the code and let me know of my mistake.

**Comment by pranv** [comment]
The only error I can think of is that the weights had to be transposed. Anything else would lead to dimensionality errors.

Or, in VGGs case, there could be an issue with the `group` parameter. But again, of this wasn't right, this would lead to dimensionality errors as well. Again, a transpose might be in necessary.

File: keras/caffe/converters.py [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+
+            # weird old trick used in paralellizing networks
+            group = layer.convolution_param.group
+            stack_size = temp_stack_size * group
+
+            # maybe not all synapses are existant
+            weights_p = np.zeros((nb_filter, stack_size, nb_col, nb_row))
+            weights_b = np.array(blobs[1].data)
+
+            chunk_data_size = len(blobs[0].data) // group
```
Comment (pranv) [inline keras/caffe/converters.py]: Just for reference, this is how chainer does this:
```python
def _setup_convolution(self, layer):
        blobs = layer.blobs
        param = layer.convolution_param
        ksize = _get_ksize(param)
        stride = _get_stride(param)
        pad = _get_pad(param)
        num = _get_num(blobs[0])
        channels = _get_channels(blobs[0])

        n_in = channels * param.group
        n_out = num
        func = functions.Convolution2D(n_in, n_out, ksize, stride, pad,
                                       nobias=not param.bias_term)
        func.W.fill(0)

        part_size = len(blobs[0].data) // param.group
        for i in six.moves.range(param.group):
            in_slice = slice(i * n_in // param.group,
                             (i+1) * n_in // param.group)
            out_slice = slice(i * n_out // param.group,
                              (i+1) * n_out // param.group)
            w = func.W[out_slice, in_slice]

            data = numpy.array(blobs[0].data[i*part_size:(i+1)*part_size])
            w[:] = data.reshape(w.shape)

        if param.bias_term:
            func.b[:] = blobs[1].data

        setattr(self.fs, layer.name, func)
        self.forwards[layer.name] = func
        self._add_layer(layer)
```

And this is how sklearn-theano does this:
def _blob_to_ndarray(blob):
```python
    """Converts a caffe protobuf blob into an ndarray"""
    dimnames = ["num", "channels", "height", "width"]
    data = np.array(blob.data)
    shape = tuple([getattr(blob, dimname) for dimname in dimnames])
    return data.reshape(shape)
```

File: keras/caffe/converters.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+            # FLATTEN
+            model.add_node(Flatten(), name=name, input=input_layer_name)
+
+        elif layer.type == 14:
+            # INNER PRODUCT OR DENSE
+            blobs = layer.blobs
+            nb_filter, stack_size, nb_col, nb_row = blobs[0].num, blobs[0].channels, blobs[0].height, blobs[0].width
+
+            weights_p = np.array(blobs[0].data).reshape(nb_filter, stack_size, nb_col, nb_row)[0, 0, :, :].T
+            weights_b = np.array(blobs[1].data)
```
Comment (pranv) [inline keras/caffe/converters.py]: Chainer:
```python
def _setup_inner_product(self, layer):
        param = layer.inner_product_param
        bias_term = param.bias_term
        if param.axis != 1:
            raise RuntimeError(
                'Non-default axis in InnerProduct is not supported')

        blobs = layer.blobs
        width, height = _get_width(blobs[0]), _get_height(blobs[0])
        func = functions.Linear(width, height, nobias=not bias_term)
        func.W.ravel()[:] = blobs[0].data
        if bias_term:
            func.b[:] = blobs[1].data
```

sklearn-theano:
```python
 elif layer_type == "INNER_PRODUCT":
            weights = layer_blobs[0].astype(float_dtype)
            biases = layer_blobs[1].astype(float_dtype).squeeze()
            fully_connected_input = blobs[bottom_blobs[0]]
            # fc_layer = Feedforward(weights, biases, activation=None)
            fc_layer = Convolution(weights.transpose((2, 3, 0, 1)), biases,
                                   activation=None)
            fc_layer._build_expression(fully_connected_input)
            layers[layer_name] = fc_layer
            blobs[top_blobs[0]] = fc_layer.expression_
```

**Comment by pranv** [comment]
@fchollet For a measure of complexity, I think the above two repos might give an estimate.

**Comment by asampat3090** [comment]
It seems like the weights for the fully connected layers were transposed incorrectly, but otherwise the weights and biases for the convolution and fully connected layers all matched up with caffe. 

However there is still something wrong. It probably isn't the group parameter since I checked the weights against caffe's after the model was imported, but it could be an issue with pooling or relu layers (made changes [here](https://github.com/asampat3090/keras/tree/caffe)). Any other ideas? It does seem like it could be a relu issue given that the keras result has no zero values and the caffe one does...but not sure how to check it. 

Caffe features sample (correct): 
```python
array([-0.        , -0.        ,  2.7828536 , ..., -0.        ,
       -0.        ,  1.70453358], dtype=float32)
```
Keras features sample: 
```python
array([[-24.98186874,  21.50024796,  -1.37967587, ..., -28.88113213,
         -3.03378105,   5.12496424]])
```

**Comment by pranv** [comment]
@asampat3090 Thank you for that.

>It does seem like it could be a relu issue given that the keras result has no zero values and the caffe one does...but not sure how to check it.

`network.get_config()` could offer some insight.

**Comment by pranv** [comment]
@asampat3090 also, do post the code changes you made (if any)

**Comment by phreeza** [comment]
@asampat3090 Maybe you could add a test to perform the check you did automatically?

**Comment by fchollet** [comment]
> Maybe you could add a test to perform the check you did automatically?

That would be great, if it's doable. Such a check should absolutely be part of testing that Caffe import does work.

**Comment by fchollet** [comment]
@asampat3090 :

> It does seem like it could be a relu issue given that the keras result has no zero values and the caffe one does...but not sure how to check it.

What model method are you using to check your results? Dropout will only be applied in 'training' mode, i.e. using `model.train`, etc. 

Also, it would be great to have your code so we can reproduce these results and investigate.

**Comment by fchollet** [comment]
Do you guys know if the Caffe format has a documentation? Layer objects have *a lot* of attributes (I count 132), and I can't seem to find any documentation for them.

By the way, with the current PR I am failing to load a number of prototexts from Model Zoo.

**Comment by pranv** [comment]
The best way to learn about caffe layer attributes is to use python's `help()`

The prototexts on model zoo are not updated to recent protocols. The fail is from parsing it.

**Comment by fchollet** [comment]
Do you have more info about format changes (when did it occur, what did it break)? I'm talking about files taken from the master branch of the Caffe repo, I would expect them to be up to date.

**Comment by pranv** [comment]
https://github.com/BVLC/caffe/pull/1694

6 months ago this was merged. I think I said the other way around. Sorry!

**Comment by pranv** [comment]
I just tried the same experiment - removed the top fully connected layer and ran the network.
I got a all-positive 4096 vector, with L1 being 0.65.
@asampat3090, I think the error could rather be due to preprocessing not being applied.

But that doesn't explain the negative values you're getting. Do share your model definition file as soon as possible.

**Comment by pranv** [comment]
@fchollet, I found out the exact issue with BVLC models. They are in V2 of prototext. Other public models like VGG are in V1. 
Partly because I wanted to use VGGnet more than anything, the code is for V1. 

Now, there is way to support both V1 and V2.  I'll have to Write `if..else` checks everywhere and handle both versions. I'll get it ready soon.

**Comment by pranv** [comment]
@fchollet, I've committed the changes to my branch. The models in Caffe repo are supported. Not sure how to get it here though

**Comment by fchollet** [comment]
@pranv you can submit a PR to the `caffe` branch in Keras. This will be helpful since I have also local changes to `caffe` that I will commit later.

**Comment by fchollet** [comment]
I have removed ~230 lines of code. Any concerns? https://github.com/fchollet/keras/commit/8a31762d0703d704befbbfe4850079bc6223f027

I haven't tested for correctness yet, so although the conversion yields valid Keras models, it is not clear whether the weights do reproduce the original Caffe network...

By the way, I believe it would be simpler to just convert Caffe models to a Keras config dictionary, then instantiate the model from this config, instead of converting the Caffe model to an instantiated Keras model. I will look into it in the near future. Any concerns?

Also I will still be looking for a simpler way to do this step: https://github.com/fchollet/keras/blob/8a31762d0703d704befbbfe4850079bc6223f027/keras/caffe/converters.py#L29-L38

**Comment by pranv** [comment]
Based a on a quick skim:

* I think you made an `if...else` decision with `prototxt` or `caffemodel`.  As you can see from the discussion above, people would want to change `prototxt`s to obtain modified models
* I assumed `convert_weights` would be helpful in embedding something like VGG net for feature extraction
* What happened in line 37?

Overall, it was a good idea to reduce almost 2 similar functions to 1.

**Comment by pranv** [comment]
Also, I had a conversation with @szagoruyko, the author of the torch equivalent and a member of the torch team(FAIR). He thought the code was fine

**Comment by fchollet** [comment]
@asampat3090 
> It seems like the weights for the fully connected layers were transposed incorrectly, but otherwise the 
weights and biases for the convolution and fully connected layers all matched up with caffe.

Checking out your changes, what was your justification for removing the transposition for the weights of the fully connected layers? That doesn't seem right.

**Comment by asampat3090** [comment]
@fchollet sorry for the delayed response. I'm using the VGG 16 layer caffemodel and have a custom prototext file (https://gist.github.com/asampat3090/c63a6f5082bab64e8a74). As for the weights, I compared the weights in caffe to those in keras and transposed those weights because they didn't match. Unfortunately it still didn't result in the correct answer given my prototxt.

**Comment by pranv** [comment]
@asampat3090, please try with the latest code again

**Comment by pranv** [comment]
>On the C++ side the image is represented as uint8 in 0-255 while on the Python side scikit-image represents the image data as float in 0-1. For this reason the Python wrapper scales first so that it can re-map the data to 0-255 to match the precomputed mean.

From caffe issue somewhere.

**Comment by asampat3090** [comment]
Sorry for posting on the older post. @fchollet referring to our previous convo (others please see #368). I used the caffemodel file for the VGG 16 layer as given in the Model Zoo (http://www.robots.ox.ac.uk/~vgg/software/very_deep/caffe/VGG_ILSVRC_16_layers.caffemodel). 

I'm not sure why the caffe model would be in training mode - I did specify the phase to be test (or so I thought) via: `net.set_phase_test()`

@pranv oh interesting...I wonder how I can prevent that - or at the very least know what that scaling is. It seems like that may be out of my control if I am using the pycaffe wrapper then. Any ideas?

**Comment by pranv** [comment]
@asampat3090, can you divide the entire image by 255 and try?

**Comment by asampat3090** [comment]
Unfortunately that didn't work. I tried to divide the caffe input by 255, which made the caffe output change and have nearly as many non-zero values as the keras output, but the numbers were still off - also, I'm not confident that would be the correct result for caffe anyways. Alternatively I tried to multiply the keras input by 255, but that didn't change much. 

It probably does have to do with how the image is interpreted though since I've verified the weights matched up with the caffe blobs for each layer. Or I suppose the layers could not be connected up correctly but that seems unlikely. 

@pranv do you think you could try running my code and see what you get?

**Comment by fchollet** [comment]
By the way, wouldn't it be easier (less prone to issues in the test code itself) to look at much simpler networks first (eg. single FC layer, or single Conv layer)? With a single-layer network any potential error (in the conversion or in the test code) should be immediately obvious.

> Or I suppose the layers could not be connected up correctly but that seems unlikely.

You can look at the Keras network via `graph.get_config(True)`.

**Comment by pranv** [comment]
@asampat3090,  I suggested that you divide keras input by 255

**Comment by asampat3090** [comment]
@pranv I actually tried that as well, but that made it worse. @fchollet yea that's actually a good point. I'm guessing you mean using the same caffemodel file and slowly building up the prototxt one layer at a time?

**Comment by fchollet** [comment]
@asampat3090 it could be done this way... but, more cleanly: couldn't we build a few single-layer (or 2-layers) sequential Caffe models, save their weights to a `.caffemodel` file, and write unit tests for the Keras-Caffe importer based on these files?

**Comment by dasguptar** [comment]
Hi,
Could I ask what the current status of the Caffe branch is? The last activity here was almost 3 weeks ago. 
Also, wow, this is really neat work!

**Comment by pranv** [comment]
@dasguptar Thanks :)

There is a small bug in convolution layers. And unit tests have to be added. I'm sure it'll be done within this week.

**Comment by mmmikael** [comment]
I noticed a small difference between Keras and Caffe's implementations of the Dropout layer.
I made a "Caffe mode" here https://github.com/mmmikael/keras/blob/caffe_m/keras/layers/core.py#L252-L276

**Comment by phreeza** [comment]
What is the status on this?

**Comment by lireagan** [comment]
What is the status on this?

**Comment by grisaitis** [comment]
@lireagan @phreeza you might want to see https://github.com/fchollet/keras/pull/921

**Comment by lireagan** [comment]
@grisaitis Thank you
